mgr/influx: Various fixes and improvements

wido opened the pull request:
The InfluxDBClient can also re-raise a ConnectionError from the
python requests module if that was caught.

The Exception might be:

<pre>ConnectionError: ('Connection aborted.', error(104, 'Connection reset by peer'))</pre>

Catch and log this error instead of having it raised which might
cause problems further down.

Signed-off-by: Wido den Hollander <wido@42on.com>

jcsp commented:
This is an improvement on having the module stop entirely, but it would also be nice to have the module raise a health error while it is failing to contact the server.

Could be as simple as each time round the reporting loop, if it succeeded then do a set_health_checks({}), and otherwise do a set_health_checks something like

```
self.set_health_checks({
'MGR_INFLUX_SEND_FAILED': {
'severity': 'warning',
'summary': 'InfluxDB module could not contact server at ...',
'detail': []
}
})
```

wido commented:
Thanks @jcsp !

How does this look?

jcsp commented on src/pybind/mgr/influx/module.py:
```diff
@@ -205,6 +226,14 @@ def send_to_influx(self):
+                self.set_health_checks({
+                    'MGR_INFLUX_SEND_FAILED': {
+                        'severity': 'warning',
+                        'summary': 'Failed to send data to Influx due to a '
```
s/a Exception/an exception/

Maybe be a bit more specific and say "an InfluxDB client exception"

jcsp replied:
Actually after looking at 20198 as well, I think it's better for these messages not to mention exceptions and just to say "Failed to send data to InfluxDB" or similar -- the important information is that we failed to send something, the exception is just the mechanism by which we found out.

jcsp commented on src/pybind/mgr/influx/module.py:
```diff
@@ -196,6 +204,19 @@ def send_to_influx(self):
+                'MGR_INFLUX_SEND_FAILED': {
+                    'severity': 'warning',
+                    'summary': 'InfluxDB module could not contact server at '
+                               '%s:%d due to a ConnectionError' %
```
let's say "connection error" instead of ConnectionError, as this string will go in front of non-programmer eyeballs

jcsp commented on src/pybind/mgr/influx/module.py:
```diff
@@ -77,29 +77,19 @@ def get_df_stats(self):
                     "time": datetime.utcnow().isoformat() + 'Z',
                     "fields": {
-                        "value": pool['stats'][df_type],
+                        "value": pool['stats'][key],
```
If we're using pool['stats'][key] here, then let's use a `for key, value in pool['stats'].items()` in the loop and avoid the second lookup

jcsp commented:
Looks good, just very minor comments

wido commented:
Thanks! How does this look?

jcsp commented:
Thanks!  We don't have automated tests that cover the actual connection to the server, so can you confirm that you've tested this by hand?

wido commented:
@jcsp I wanted to confirm, but I couldn't. After lunch I noticed that the Influx module broke.

'percent_used' by the 'df' is a string and not a float.

<pre>{'wr': 202L, 'quota_objects': 0L, 'bytes_used': 210763776L, 'max_avail': 8784400384L, 'rd': 0L, 'rd_bytes': 0L, 'objects': 201L, 'percent_used': '2.34', 'kb_used': 205824L, 'quota_bytes': 0L, 'raw_bytes_used': 632291328L, 'wr_bytes': 211812352L, 'dirty': 201L}</pre>

This caused InfluxDB to complain as the value should be a int or a float.

For now I've reverted the commit partially and just added the fields.

Mainly the *rd* and *wr* values were missing as they are important to see how many IOps a pool is doing.

I wanted to stay away from a try/except block with all kinds of casts.

Can we fix that the Mgr gives 'df' back with that value as a float?

wido commented:
The error handling is working though:

<pre>root@alpha:~# ceph influx config-set port 8080
Configuration option port updated
root@alpha:~# ceph -s
  cluster:
    id:     6e4daf97-c6a1-47c9-9606-6dd92c527cd2
    health: HEALTH_WARN
            Failed to send data to InfluxDB server at delta:8080 due to an connection error
 
  services:
    mon: 3 daemons, quorum alpha,bravo,charlie
    mgr: alpha(active)
    osd: 3 osds: 3 up, 3 in
 
  data:
    pools:   1 pools, 64 pgs
    objects: 201 objects, 201 MB
    usage:   3763 MB used, 26653 MB / 30416 MB avail
    pgs:     64 active+clean
 
root@alpha:~# ceph influx config-set port 8086
Configuration option port updated
root@alpha:~# ceph health
HEALTH_OK
root@alpha:~#</pre>

The current code works and has been tested on my local system.

jcsp commented:
> Can we fix that the Mgr gives 'df' back with that value as a float?

I think we should change that, yeah.  @jecluis added that field in 159dc9ae4e as a string: Joao, do you have strong feelings about changing that field to a floating point number so that readers don't need to parse it?

wido commented:
Is this one ready to be merged?
